### PR TITLE
Show detail book

### PR DESCRIPTION
**Pull request by NguyenThiHangB**

### CODE REVIEW

**Comment by NguyenThiHangB** [comment]
ready

**Comment by NguyenThiHangB** [comment]
ready

File: app/controllers/books_controller.rb [hunk 1]
```diff
@@ -1,13 +1,21 @@
 class BooksController < ApplicationController
+  before_action :find_book, only: %i(show edit update destroy)
```
Comment (mdmahmudurrahman) [inline app/controllers/books_controller.rb]: 1. You can rename the filter like as "load_book"
2. Use **except** if the number of actions are more than 3

File: app/controllers/books_controller.rb [hunk 2]
```diff
@@ -1,13 +1,21 @@
...
+
   private
 
   def book_params
     params.require(:book).permit :name, :description, :image,
       :publish_date, :price, :category_id
   end
+
+  def find_book
+    @book = Book.find_by id: params[:id]
```
Comment (mdmahmudurrahman) [inline app/controllers/books_controller.rb]: What will happen if it do not find the @book. I meant how you can handle the case if @book is nil?

File: app/controllers/categories_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+class CategoriesController < ApplicationController
+  before_action :find_category, only: %i(create edit update destroy)
```
Comment (mdmahmudurrahman) [inline app/controllers/categories_controller.rb]: The same here like as the comment in **BooksController**

File: app/controllers/categories_controller.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  def destroy; end
+
+  private
+
+  def category_params
+    params.require(:category).permit :title, :description
+  end
+
+  def find_category
+    @category = Category.find_by id: params[:id]
```
Comment (mdmahmudurrahman) [inline app/controllers/categories_controller.rb]: The same here like as the comment in BooksController

File: app/models/category.rb [hunk 5]
```diff
@@ -1,3 +1,6 @@
 class Category < ApplicationRecord
   has_many :books
+
+  scope :index_category, ->{select("id, title, description").
+    order created_at: :desc}
```
Comment (mdmahmudurrahman) [inline app/models/category.rb]: Rename it like as **latest** as you are getting the records in the descending order of **created_at**

File: app/views/categories/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+              Settings.user.per_page %>
+          </td>
+          <td><%= link_to category.title, category %></td>
+          <td><%= link_to category.description, category %></td>
+          <td>
+            <%= link_to t(".edit"), edit_category_path(category), class: "btn btn-success" %>
+          </td>
+          <td>
+            <%= link_to t(".delete"), category, method: :delete,
+            data: { confirm: t(".confirm") }, class: "btn btn-danger"%>
```
Comment (mdmahmudurrahman) [inline app/views/categories/index.html.erb]: Please add 2 space indentation as you broke the line

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by NguyenThiHangB** [comment]
ready

**Comment by NguyenThiHangB** [comment]
ready
Link task: https://dev.framgia.com/issues/185251

File: app/controllers/categories_controller.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+class CategoriesController < ApplicationController
+  before_action :load_category, except: %i(index new create)
+  def index
```
Comment (mdmahmudurrahman) [inline app/controllers/categories_controller.rb]: After before_action, please add a new line

File: app/controllers/categories_controller.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    @categories = Category.list_category
+                          .page(params[:page])
+                          .per Settings.per_page
+  end
+
+  def new
+    @category = Category.new
+  end
+
+  def create
```
Comment (mdmahmudurrahman) [inline app/controllers/categories_controller.rb]: All type of user can create category, right? (?)

File: app/views/categories/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,33 @@
+<% provide :title, t(".categories") %>
+<h1><%= t ".categories" %></h1>
+<%= paginate @categories, theme: "twitter-bootstrap-3"%>
+<table class="table table-bordered">
+    <thead>
```
Comment (mdmahmudurrahman) [inline app/views/categories/index.html.erb]: Indentation

File: app/views/categories/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+              Settings.user.per_page %>
+          </td>
+          <td><%= link_to category.title, category %></td>
+          <td><%= link_to category.description, category %></td>
+          <td>
+            <%= link_to t(".edit"), edit_category_path(category), class: "btn btn-success" %>
+          </td>
+          <td>
+            <%= link_to t(".delete"), category, method: :delete,
+            data: {confirm: t(".confirm")}, class: "btn btn-danger"%>
```
Comment (mdmahmudurrahman) [inline app/views/categories/index.html.erb]: Add 2 space indentation as you break a line here

File: app/views/categories/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+      <% @categories.each_with_index do |category, index| %>
+        <tr>
+          <td>
+            <%= create_index params[:page], index,
+              Settings.user.per_page %>
+          </td>
+          <td><%= link_to category.title, category %></td>
+          <td><%= link_to category.description, category %></td>
+          <td>
+            <%= link_to t(".edit"), edit_category_path(category), class: "btn btn-success" %>
```
Comment (mdmahmudurrahman) [inline app/views/categories/index.html.erb]: Any type of user can edit a category, right?

File: app/controllers/categories_controller.rb [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      redirect_to categories_path
+    else
+      flash[:danger] = t ".danger"
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update; end
```
Comment (mdmahmudurrahman) [inline app/controllers/categories_controller.rb]: Why update action body is empty ?

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by NguyenThiHangB** [comment]
ready
Link task: https://dev.framgia.com/issues/185251

**Comment by mdmahmudurrahman** [comment]
(commented)
Please answer the question. Any type of user can create a category (?)

**Comment by mdmahmudurrahman** [comment]
commented
Please answer the question. Any type of user can create a category (?)

File: app/controllers/categories_controller.rb (hunk 8 above)
Comment (NguyenThiHangB) [reply app/controllers/categories_controller.rb]: Only admin can manager category. I fixed it

**Comment by NguyenThiHangB** [comment]
ready

**Comment by mdmahmudurrahman** [comment]
commented
This (pr) can be (merged)

**Pull request merged**